### ISSUE TITLE
Support Python3 in ViewVC

### ISSUE THREAD

**Issue by 0-wiz-0** [viewvc/viewvc#138]
When trying to install using python3.6, I see:
````
/usr/pkg/bin/python3.6  viewvc-install --prefix=/usr/pkg/share/viewvc --destdir=/scratch/www/viewvc/work/.destdir)
  File "viewvc-install", line 50
    ("bin/cgi/viewvc.cgi",        "bin/cgi/viewvc.cgi",        0755, 1, 0, 0),
                                                                  ^
SyntaxError: invalid token
*** Error code 1
````
Please add support for python3 here, and in other places, if necessary. Thanks!

**Comment by cmpilato** [comment]
If you are planning to use ViewVC as a CVS viewer only, Python3 support is a reasonable request.  And certainly for as long as Python3 has been generally available, it's not like we're talking about riding the bleeding edge here. :-)  Unfortunately, Subversion's own Python language bindings are, unless I'm mistaken, currently locked on Python2.  Any suggestions how to best deal with this?

**Comment by 0-wiz-0** [comment]
There are alternative subversion bindings for python that support Python 3.x:
https://github.com/jelmer/subvertpy

**Comment by cmpilato** [comment]
Sure, but `subvertpy` isn't a drop-in replacement for the Apache Subversion project's own language bindings.

Again, I'm not saying that this request isn't a good one.  I mean only to establish the scope.  At a minimum, you're talking about rewriting all of ViewVC's core code and supporting tools (viewvc-install, make-database, the CVS/SVN commit hook scripts, cvsdbadmin, svndbadmin, the CGI/WSGI wrappers, etc.) to support Python3 directly and use Python3-compatible third-party modules (Subversion, pygments, ezt, etc.).

**Comment by futatuki** [comment]
Hi, I'm interested in this issue, and cannot wait for Subversion's swig Python binding support Python 3, so I've implemented alternative subversion accsess sub-package for vclib, vclib.altsvn,  based on vclib.svn with bridge code to C API, written in Cython. 

https://github.com/futatuki/viewvc/tree/cython-svnlib
(document: https://github.com/futatuki/viewvc/blob/cython-svnlib/src/lib/README-altsvn.txt)

In this module, strings other than paths treat as str (decoded str on Python 3 and same as bytes on Python 2), and  paths are treat as bytes.

I think It is not enough quality to make a pull request yet, but if someone interested in my work, I'm willing to make it a pull request.

**Comment by futatuki** [comment]
viewvc-install, bin/make-database, bin/svndbadmin is now work on Python 3.7 (due to syntax compatibility, etc., dropping support Python 2.5 and earlier).
https://github.com/futatuki/viewvc/commit/86266db10a0dd245aada20897ae256171fd011bf

Of course, it is only small part and no mean if core part of ViewVC is not support Python 3.x. 

The master branch seems to keep supporting Python 2.4 and 2.5, so I'll continue to work on my own branch. If the branch to support Python 3.x on viewvc repo is created, I'll make a pull request.

**Comment by futatuki** [comment]
> In this module, strings other than paths treat as str (decoded str on Python 3 and same as bytes on Python 2), and paths are treat as bytes.

On the way of modification viewvc.py to support Python 3, I realize this policy is not good. There exists many places to handle pathes and path parts as str, from PATH_INFO, config options, parts of URL, etc. Perhaps it is better that pathes are also treated as str except where it is really need to treat them as bytes (i.e. arround Subversion C API, etc.), though some functions for svn repos works on standalone server on Python 3.7.

Any suggestions ?

**Comment by cmpilato** [comment]
@futatuki I'm intrigued by the work you're doing here, and thoroughly impressed that you've tackled a Python3-based implementation of the Subversion API bindings!  #rockstar 

I can't help but wonder, though, if your efforts would be more generally valuable if you applied them toward helping the Apache Subversion development community provide Python3 versions of the official Subversion Python bindings.  I confess that I don't know all that's involved there, but my recollection is that there's a little bit of manually-maintained glue code that would need to be ported, plus portion the Swig `.i` interface files.  http://www.swig.org/Doc1.3/Python.html#Python_python3support indicates that `FILE *` typemaps are broken, which looks like might have a very limited affect on Subversion bindings (because Subversion mostly uses the APR `apr_file_t` abstraction for I/O).

Given that, I would be seriously tempted to just move ViewVC `master` wholesale to Python3.  It looks like some of the major modules that ViewVC depends on (`chardet`, `pygments`, etc.) have Python3 ports in place already.

**Comment by futatuki** [comment]
I think that writing generic wrapper API is so much difficult than writing brigding code for specific purpose, that is why I choose the later. Especially APR pool life time management behind Python object makes it complex for Subversion C API.  ... and it is too difficult for me to understand what swig Python bindings code do, (though I understood a bit by reading glue code for implementing bridge module).

> Given that, I would be seriously tempted to just move ViewVC master wholesale to Python3. It looks like some of the major modules that ViewVC depends on (chardet, pygments, etc.) have Python3 ports in place already.

Actually, view_diff, view_directory, view_log, view_markup, view_annotate, view_diff, view_checkout, download_tarball functions for Subversion local repository now work on Python 2.7/3.6/3.7 on the same code base with my altsvn module (of course, there can be oversights and/or bugs).  MySQLdb itself seems to be no support for Python 3, but its fork mysqlclient-python supports Python 3.  ViewVC's version of ezt didn't support Python 3, but I've imported https://github.com/gstein/ezt with https://github.com/gstein/ezt/issues/5, and merge feature to allow callback object for value to write, then it seems to work. So I think there is nothing to block ViewVC to support Python 3 except Subversion Python bindings :)

**Comment by futatuki** [comment]
Now it works for CVS repos also on Python 3.7, if not using rcsparse module (option `use_rcsparse=0`), on my branch https://github.com/futatuki/viewvc/tree/python3_support_with_altsvn.

However as using `rcsparse` on Python 3 have somewhat difficulty for bytes/str context distinguishment, it is not work for binary file checkout (i.e. repos.openfile()) yet.

**Comment by cmpilato** [comment]
I wonder (aloud): how hard might it be to make an SVN vclib implementation that uses the Subversion command-line tool, possibly with its `--xml` output mode?

**Comment by futatuki** [comment]
> However  as using `rcsparse` on Python 3 have somewhat difficulty for bytes/str context distinguishment, it is not work for binary file checkout (i.e. repos.openfile()) yet.

Now It also work with `use_rcsparse=1` including binary file checkout composed by binary diff, except the problem of charset/encoding of commit info, like author, log message, and so on (yes, it is also problem if using `bincvs` and perhaps current ViewVC didn't special care for it, but it is another issue).

**Comment by futatuki** [comment]
> I wonder (aloud): how hard might it be to make an SVN vclib implementation that uses the Subversion command-line tool, possibly with its `--xml` output mode? Could `vclib/svn/svn_cmdline.py` become a thing?

I've tried to implement a bit.  xml parser included in `xml.etree.ElementTree` standard library seems to work fine.  And with svn 1.9 and above, `--show-item=ARG` with `--no-newline` can be used for simple output.  `CmdLineSubversionRepository.__init__()`, `open()`,  `itemtype()` are implemented, and as a results of refactoring,  only 13 public methods remain to be unimplemented :) (Yes, of course I know I only implement easier functions, so it is after this what is hard to implement.)

svn_commandline.py:
https://github.com/futatuki/viewvc/blob/vclib_svn_commandline/lib/vclib/svn/svn_commandline.py

(branch, with refactoring for lib/vclib/svn subpackage)
https://github.com/futatuki/viewvc/tree/vclib_svn_commandline

**Comment by futatuki** [comment]
Putting aside its performance, now commandline client `svn` based driver `svn_commandline.py` works on Python 2.7. (not tested on Python 3 yet).

**Comment by futatuki** [comment]
Even with PR #185,  `bin/svndbadmin` still needs swig Python bindings directly.  To support Python 3.x with 'pure script' approach allowing without official Python binding,  I think it is goot to extend `*SubversionRepository` classes in `vclib.svn` to support functions needed by `bin/svndbadmin`.

**Comment by futatuki** [comment]
Now I posted patches for swig Python bindings for Python 3 support against swig-py3 branch of Subversion repo to Subversion development mail list.   With them, I've succed to make ViewVC work on Python 3 with swig Python bindings.

However, I also found an issue on swig-py3 which seems to need  fix (or some kind of decision),  so I don't think official release of it will be soon.

**Comment by futatuki** [comment]
just for our information. 

@jun66j5 suggests that even meta data of Subversion may includes strings with incorrect sequence as UTF-8 (https://trac.edgewall.org/ticket/4321, https://trac.edgewall.org/browser/tags/trac-1.2.3/tracopt/versioncontrol/svn/svn_fs.py?marks=984-986#L971) so it migtht be needed to use bytes for interface between Subversion C API and Python 3 in most case. 

On the other hand, core part of ViewVC  (viewvc.py) expects str for repos API,  so it may be need to care within vclib.svn modules on Python 3.

**Comment by futatuki** [comment]
Now I make a branch [python3_support_without_altsvn](https://github.com/futatuki/viewvc/tree/python3_support_without_altsvn) from python3_support_with_altsvn branch, by excluding changes for altsvn binary module and refactoring of lib/vclib. Though It is not tested and clean up, I'll do it and make it to PR(s).

**Comment by futatuki** [comment]
Note:  The Python 3 support code for Subversion in python3_support_without_altsvn branch described above is incomplete, because svn.* module API has changed after I wrote it.

**Comment by cmpilato** [comment]
@futatuki I've taken the liberty of merging your `python3_support_without_altsvn` branch into this repository's `python3-support-futatuki` branch.  I'll work to catch that branch up with the latest changes made to ViewVC since the branch divergence.

**Comment by futatuki** [comment]
@cmpilato than you!  Now I also started to work on `python3-support-futatuki` branch.

**Comment by cmpilato** [comment]
@futatuki I realized yesterday that I acted prematurely in the past when I branched off `1.2.x` and made `master` target 1.3.0 without ever have release anything in 1.2.  So `master` is once again aimed at 1.2.0 and I intend to release ASAP a 1.2.0 that has all the features and fixes collected over the past 10 years that have been sitting in `master`.  Once this is done, I'll re-create the `1.2.x` maintenance branch and `master` will be ready for 1.3.0-dev and for Python 3.  I'm really sorry for the confusion and the code churn, but I realized that there were a bunch of changes I needed to make to get 1.2.0 ready and I was going to have to try to make those changes in both the 1.2.x line and the 1.3.x line to avoid regression.  It's just a really poor development model.

It sounds like you're suggesting that the minimum Python we support at that time be Python 3.3  I'm okay with that.  After all, 3.4 is already out of support!

**Comment by cmpilato** [comment]
Okay.  I have release ViewVC 1.2.0 and re-initiated the `1.2.x` release branch.  So I've also once again made `master` be the place where 1.3.0 development can take place.  I took a stab at rebasing the `python3-support-fuatuki` on the new `master`, but there were conflicts I wasn't quite sure how to resolve.  I'll try to take another look at it before trying an alternate approach.

**Comment by futatuki** [comment]
I've made a new branch [branch-fix-python3-support-futatuki](https://github.com/futatuki/viewvc/tree/branch-fix-python3-support-futatuki) from `viewvc/python3-support-futatuki`, and rebased on branch point where `old-master` had branched (dd8d8aa1e5ed7a97c2b10c5bed47d8fc52738568).
This branch already merged `viewvc/master` branch (01de73291208c2c15586433087f8bb5695b0d713) and my local fix,  however it still doesn't work for svn repos because of bytes/str mixture, so I don't make a PR for `master` branch (yet).

**Comment by futatuki** [comment]
Now all commits in `futatuki/branch-fix-python3-support-futatuki` branch are in `viewvc/python3-support` branch. So I drop my `futatuki/branch-fix-python3-support-futatuki` and I've started to  work on `viewvc/python3-support` branch.

**Comment by futatuki** [comment]
As I wrote in [comment above](https://github.com/viewvc/viewvc/issues/138#issuecomment-441966007) and [comment for #165[(https://github.com/viewvc/viewvc/issues/165#issuecomment-438725783), commit log message and authore can be other than 'ascii' or 'UTF-8', both of CVS and Subversion. 

How do we deal this between viewvc.py and vclib? 

(a) `viewvc.py` passes encoding as hint, and `vclib` returns log message and author as `str` by using encoding hint.
(b) `vclib` return log message and author as `bytes`, and then `viewvc.py` decode it

Current implementation of `vclib/ccvs/bincvs.py` and `vclib/ccvs/rcspurts/*` return them as `str` but don't care encoding, and `vclib/svn/*` return them as `bytes` (by change on Subversion SWIG python binding interface after I modified `vclib/svn/*` for Python 3).

**Comment by cmpilato** [comment]
I think (a) is the only reasonable solution.  The `vclib/` interface needs to have a clear and clean promise -- that it operates in `str` types for all things semantically intended to be strings (paths and metadata/properties).  And yes, if we need to pass a hint into that layer from above, so be it.

**Comment by futatuki** [comment]
To tell the truth, I've already started to modify source files in accodance with policy (a) :-)

However property values in Subversion repository and file contents can be binary data, so we should treat it as `bytes`, I think.  For properties,  we use `is_undisplayable()` to determine they can be displayed or not (function `is_undisplayable()` on python3-support branch is currently broken -- it always returns 0, though).

**Comment by futatuki** [comment]
Now `lib/vclib/svn/svn_repos.py` and `lib/vclib/svn/svn_ra.py` works for Python 3 on python3-support branch, at least level. (Though query view does not work yet)

**Comment by cmpilato** [comment]
As for the properties, I still think that the vclib interface should demand String property names and values.  This probably means changing the `itemprops()` interface so that *it* makes the decision about what's "undisplayable".  That interface today returns a dictionary mapping property names to values.  Let's just make it such that names are always `str`, values are either `str` or `bytes`.  Callers can decide for themselves what to do about that.  In our case, it'll be to turn every property with a `bytes` value into an `_item()` with `undisplayable` set, right?

I'll tackle this today.

**Comment by cmpilato** [comment]
Finished in 8d37dd68.  Something similar will need to be done with the revision properties.

**Comment by cmpilato** [comment]
Okay, I made similar changes to the `revinfo()` interface in 5bcca36b.

**Comment by cmpilato** [comment]
@futatuki Are you tracking anywhere a list of what remains to be done on this branch?  I'd like to continue contributing, but don't want to overlap with whatever you are doing.

**Comment by futatuki** [comment]
Here is a list to do.

1. Add per repository config parameter to specify encoding hint.

2. Check bin/make-databse again. 
This one is not checked yet, on this branch. (It may also rework arround subproces.Popen vs popen.popen)
3. Update bin/svndbadmin. 
This one was rewrited not to use svn.* modules directly, by changing vclib Repository class 
abstruction in my python3_support_with_altsvn branch, so it is not merged to python3-support branch.
4. Check bin/cvsdbadmin.
This might work on my branch and merged straight forward, so I think it will also work on this branch, but not checked yet.
5. Fix cvsdb query feature (in lib/viewvc.py).
This is not work on this branch yet. (I can't remember whether it worked with Python 3 on my branch or not.)
6.  Check bin/loginfo-handler.
This isn't touched except by using 2to3.
7. Check http server interfaces other than bin/standalone.py

I need to rebuild MySQL/Mariadb environment for 2.-6. , so it is stuck a bit.

**Comment by cmpilato** [comment]
I can pick up the first item (per-repository encoding hint).

**Comment by cmpilato** [comment]
> 1. Add per repository config parameter to specify encoding hint.

This is done.

> 2. Check bin/make-databse again.
>    This one is not checked yet, on this branch. (It may also rework arround subproces.Popen vs popen.popen)

`make-database` is unique in that it actually wants to feed data into stdin of the process.  `popen.popen()` is really for reading data from a pipe only.  So I'm okay with make-database doing it's own thing.  I've been using a Docker stack to play with ViewVC and Subversion+PY3 and MariaDB.  I tweaked `make-database` a bit today to require `--force` if the database already exists.  But it seems to otherwise work for me.

> 3. Update bin/svndbadmin.
>    This one was rewrited not to use svn.* modules directly, by changing vclib Repository class
>    abstruction in my python3_support_with_altsvn branch, so it is not merged to python3-support branch.

When I run `svndbadmin`, I get some really ugly-looking errors that I don't currently understand:
```
MySQLdb._exceptions.OperationalError: (1267, "Illegal mix of collations (latin1_bin,IMPLICIT) and (utf8_general_ci,COERCIBLE) for operation '='")
```

> 7. Check http server interfaces other than bin/standalone.py

I'm getting some irritating errors with bin/standalone.py, too.  Nearly ever request throws an Exception:

```
----------------------------------------
Exception happened during processing of request from ('172.28.0.1', 51950)
Traceback (most recent call last):
  File "/usr/lib64/python3.6/socketserver.py", line 320, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/usr/lib64/python3.6/socketserver.py", line 351, in process_request
    self.finish_request(request, client_address)
  File "/usr/lib64/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib64/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/usr/lib64/python3.6/http/server.py", line 418, in handle
    self.handle_one_request()
  File "/usr/lib64/python3.6/http/server.py", line 407, in handle_one_request
    self.wfile.flush() #actually send the response if not already done.
ValueError: I/O operation on closed file.
----------------------------------------
```

**Comment by cmpilato** [comment]
@futatuki Given that the Subversion project is posting release candidates for 1.14.0 now, should we merge `python3-support` back into `master` and continue our work there?  I kinda think so, but am interested in your thoughts.

**Comment by futatuki** [comment]
I worried about a kind of degrade by merging incomplete code, however I realize that if continue to work on python3-support,  changes for feature or bug fix not related about Python 3 poring make extra cost.  So I agree to merge.

**Comment by futatuki** [comment]
> When I run svndbadmin, I get some really ugly-looking errors that I don't currently understand:
> 
> `MySQLdb._exceptions.OperationalError: (1267, "Illegal mix of collations (latin1_bin,IMPLICIT) and (utf8_general_ci,COERCIBLE) for operation '='")

It seems this is caused by comparison of varchar type column which is specified '`binary`' charset  and 'utf-8' string which is provided by our SQL statement.

I specified 'utf-8' as client charset for MySQLdb.connection(), but it seems it was incorrect, because we specified 'binary' for charset of varchar columns.

**Comment by cmpilato** [comment]
> I worried about a kind of degrade by merging incomplete code, however I realize that if continue to work on python3-support, changes for feature or bug fix not related about Python 3 porting make extra cost. So I agree to merge.

Merged!

**Comment by cmpilato** [comment]
> I specified 'utf-8' as client charset for MySQLdb.connection(), but it seems it was incorrect, because we specified 'binary' for charset of varchar columns.

History lesson:  ViewVC inherited its database schema from the Bonsai project.  You could use Bonsai and ViewVC side-by-side, pointing at the same database, for different ways to access your CVS commit history.  But Bonsai is dead -- four years ago all its open issue were RESOLVED WONTFIX with a notice about it being decommissioned.  I think the reality of Bonsai compatibility in ViewVC 1.3.0 is that there were be none.

It's time for ViewVC to move forward with a new schema.  Ideally, we need to be able to migrate data for folks with massive commits databases to the new schema, though I truly doubt that's possible merely with MySQL's tool chain.  But if we're getting a new schema, it sure would be nice to design one that's truly suited to Subversion (supporting directory-only versions, etc.), not just for CVS.  We're never get another opportunity like this one to do so.  And that's a _whole_ different and large task...

**Comment by futatuki** [comment]
> > 3. Update bin/svndbadmin.
> >    This one was rewrited not to use svn.* modules directly, by changing vclib Repository class
> >    abstruction in my python3_support_with_altsvn branch, so it is not merged to python3-support branch.
> 
> When I run `svndbadmin`, I get some really ugly-looking errors that I don't currently understand:
> 
> ```
> MySQLdb._exceptions.OperationalError: (1267, "Illegal mix of collations (latin1_bin,IMPLICIT) and (utf8_general_ci,COERCIBLE) for operation '='")
> ```
> 
> > 7. Check http server interfaces other than bin/standalone.py
> 
> I'm getting some irritating errors with bin/standalone.py, too. Nearly ever request throws an Exception:
> 
> ```
> ----------------------------------------
> Exception happened during processing of request from ('172.28.0.1', 51950)
> Traceback (most recent call last):
>   File "/usr/lib64/python3.6/socketserver.py", line 320, in _handle_request_noblock
>     self.process_request(request, client_address)
>   File "/usr/lib64/python3.6/socketserver.py", line 351, in process_request
>     self.finish_request(request, client_address)
>   File "/usr/lib64/python3.6/socketserver.py", line 364, in finish_request
>     self.RequestHandlerClass(request, client_address, self)
>   File "/usr/lib64/python3.6/socketserver.py", line 724, in __init__
>     self.handle()
>   File "/usr/lib64/python3.6/http/server.py", line 418, in handle
>     self.handle_one_request()
>   File "/usr/lib64/python3.6/http/server.py", line 407, in handle_one_request
>     self.wfile.flush() #actually send the response if not already done.
> ValueError: I/O operation on closed file.
> ----------------------------------------
> ```

Then, let's make them new issues on master, shall we? (I've checked where `request.wfile` is closed for the latter problem, but it is not clue, yet)

**Comment by cmpilato** [comment]
I agree.  Let's close this issue out and take up individual ones from here
on out.

>

**Comment by cmpilato** [comment]
Closing this is essentially complete.  `master` (1.3.0-to-be) has been converted to Python 3, and while there remain some dangling issues there (not least of which is that Subversion still has not quite released their own py3 support), we can take those up individually.
